Docs: Externally shared dashboard datasource plugin support (#1496)
grafana/plugin-tools: docusaurus/docs/how-to-guides/data-source-plugins/add-support-for-externally-shared-dashboards.md

Because of this, it's necessary to not pass any frontend-transformed body to the request, as it won't be used in the externally shared dashboard panel request.

:::note

Frontend data sources are not compatible with externally shared dashboards.
To convert a frontend data source plugin into a backend plugin, refer to
[convert a frontend data source to backend](./convert-a-frontend-datasource-to-backend).

:::

## Support externally shared dashboards in your data source plugin

To make your data source plugin work in an externally shared dashboard scope, follow these steps:

1.  Extend your DataSource class from `DataSourceWithBackend`

    ```ts
    export class MyDataSourceClass extends DataSourceWithBackend<TQuery, TOptions> {
      // your logic
    }
    ```

2.  Implement the `query` method with your customized code, if necessary. Don't transform the request body if this will change the backend query response (targets property). This body won't be passed as argument when calling the shared externally dashboard endpoint.

> 3.  In the `query` method, call `super.query(request)`.

Reviewer: I struggle with the layout here as 2 is optional, and 3 is only required if you do 2. It doesn't feel like a 1-4 todo list

Author: you're right. I can merge them in a single point

Author: @sympatheticmoose check again. I also modified the example, so the developer understands that the response can be manipulated

Author: hey @sympatheticmoose . were you able to check the changes?

---
id: add-support-for-externally-shared-dashboards
title: Add support for externally shared dashboards
description: How to add support for externally share dashboards (previously called Public dashboards).
keywords:
  - grafana
  - plugins
  - plugin
  - externally shared dashboards
  - public dashboards
  - data source
  - datasource
---

> Externally shared dashboards (previously called Public dashboards) retrieve the query from the backend, instead of receiving it from the frontend. This is to avoid exposing sensitive data and performing unauthorized queries.

Reviewer: ```suggestion
[Externally shared dashboards](https://grafana.com/docs/grafana/latest/dashboards/share-dashboards-panels/shared-dashboards/#externally-shared-dashboards) (previously called Public dashboards) allow Grafana users to share access to their dashboards with anyone, without needing to add them to their Grafana organization as a user. When a dashboard is accessed in this way, it retrieves the query from the backend, instead of receiving it from the frontend. This is to avoid exposing sensitive data and performing unauthorized queries.
```

It might help to provide a link back to the main docs and briefly describe the feature.

---
id: add-support-for-externally-shared-dashboards
title: Add support for externally shared dashboards
description: How to add support for externally share dashboards (previously called Public dashboards).
keywords:
  - grafana
  - plugins
  - plugin
  - externally shared dashboards
  - public dashboards
  - data source
  - datasource
---

> Externally shared dashboards (previously called Public dashboards) retrieve the query from the backend, instead of receiving it from the frontend. That is due to security purposes, to not expose sensitive data and also not be able to perform unauthorized queries.

Reviewer 2: ```suggestion
Externally shared dashboards (previously called Public dashboards) retrieve a query from the backend, instead of receiving it from the frontend. This is to avoid exposing sensitive data and performing unauthorized queries.
```

Author: I would leave `**the** query`, as it's the panel query to be performed

> Because of that, it is necessary to not pass any frontend-transformed body to the request, as it will not be used in the externally shared dashboard panel request.

Reviewer 2: Do you mean that frontend-transformed bodies are automatically not passed to the request or that users need to avoid doing this?

```suggestion
Because of this, it's necessary to not pass any frontend-transformed body to the request, as it won't be used in the externally shared dashboard panel request.
```

:::note

Frontend data sources are not compatible with externally shared dashboards.

> To convert a frontend data source plugin into a backend plugin, see

Reviewer 2: ```suggestion
To convert a frontend data source plugin into a backend plugin, refer to
```

> [convert a frontend data source to backend](https://grafana.com/developers/plugin-tools/how-to-guides/data-source-plugins/convert-a-frontend-datasource-to-backend).

Reviewer 2: ```suggestion
[convert a frontend data source to backend](/convert-a-frontend-datasource-to-backend).
```

:::

> ## Support externally shared dashboards feature in your data source plugin

Reviewer 2: ```suggestion
## Support externally shared dashboards in your data source plugin
```

> To make your data source plugin work in an externally shared dashboard scope, it is necessary to:

Reviewer 2: ```suggestion
To make your data source plugin work in an externally shared dashboard scope, follow these steps:
```

> 1.  Extend your DataSource class from `DataSourceWithBackend`

Reviewer 2: ```suggestion
1.  Extend your `DataSourceClass` from `DataSourceWithBackend`
```

    ```ts
    export class MyDataSourceClass extends DataSourceWithBackend<TQuery, TOptions> {
      // your logic
    }
    ```

2.  Implement the `query` method with your customized code, if necessary.

> 3.  In the `query` method, do not transform the request body if this will change the backend query response (targets property). This body will not be passed as argument when calling shared externally dashboard endpoint.

Reviewer 2: This should be text under step 2 instead of it's own step since it's not telling the user to do anything, but something to keep in mind as they implement the method in step 2.
```suggestion
3.  In the `query` method, don't transform the request body if this will change the backend query response (targets property). This body won't be passed as argument when calling the shared externally dashboard endpoint.
```

> 4.  Then, in the `query` method, call `super.query(request)`.

Reviewer 2: ```suggestion
4.  In the `query` method, call `super.query(request)`.
```

>     That's where externally shared dashboard endpoint is called.

Reviewer 2: ```suggestion
    This is where the externally shared dashboard endpoint is called.
```
